### ISSUE TITLE
Using cond! together with count! unable to infer enough type information about `_`

### ISSUE THREAD

**Issue by gz** [Geal/nom#302]
Not sure if it's a bug in my code or in the library but if I'm trying to parse something using the following reduced example:

```
fn parse_branch_entries(input: &[u8]) -> IResult<&[u8], Option<Vec<u64>> > {
    chain!(input,
        entries: cond!(true, count!(le_u64, 3)),
        || { entries }
    )
}
```

it tells me:

```
error[E0282]: unable to infer enough type information about `_`
   --> src/parser/mod.rs:394:5
    |
394 |     chain!(input,
    |     ^ cannot infer type for `_`
    |
    = note: type annotations or generic parameter binding required
    = note: this error originates in a macro from the standard library
```

**Comment by Geal** [comment]
this is due to type inference issues. if you isolate `cond!(true, count!(le_u64, 3))` in its own function, it might get fixed.

**Comment by Geal** [comment]
Hi, this should be fixed in df5629e
